proposals/ref-struct-interfaces.md

```diff
@@ -147,61 +151,113 @@ The new rules apply to the `heap` constraint:
 
 ### Binary breaking change
-Adding `allow T: ref struct` to an existing API is not a source breaking change. It is purely expanding the set of allowed types for an API. Need to track down if this is a binary breaking change or not. Unclear if updating the attributes of a generic parameter constitute a binary breaking change.
+
+Adding `allows ref struct` to an existing API is not a source breaking change. It is purely expanding the set of allowed types for an API. Need to track down if this is a binary breaking change or not. Unclear if updating the attributes of a generic parameter constitute a binary breaking change.
 
 ## Considerations
 
 ### Runtime support
+
 This feature requires several pieces of support from the runtime / libraries team:
+
 - Preventing default interface methods from applying to `ref struct`
-- API in `System.Reflection.Metadata` for encoding the `gpAcceptByRefLike` value.
+- API in `System.Reflection.Metadata` for encoding the `gpAcceptByRefLike` value
 - Support for generic parameters being a `ref struct`
 
-Most of this support is likely already in place. The general `ref struct` as generic parameter support is already implemented as described [here](https://github.com/dotnet/runtime/blob/main/docs/design/features/byreflike-generics.md). It's possible the DIM implementation already account for `ref struct`. But each of these items needs to be tracked down.
+Most of this support is likely already in place. The general `ref struct` as generic parameter support is already implemented as described [here][byref-like-generics]. It's possible the DIM implementation already account for `ref struct`. But each of these items needs to be tracked down.
+
+## API versioning
+
+The `allows ref struct` anti-constraint can be safely applied to a large number of generic definitions that do not have implementations. That means most delegates, interfaces and `abstract` methods can safely apply `allows ref struct` to their parameters. These are just API definitions without implementations and hence expanding the set of allowed types is only going to result in errors if they're used as type arguments where `ref struct` are not allowed.
+
+API owners can rely on a simple rule of "if it compiles, it's safe". The compiler will error on any unsafe uses of `allows ref struct`, just as it does for other `ref struct` uses.
+
+At the same time though there are versioning considerations API authors should consider. Essentially API owners should avoid adding `allows ref struct` to type parameters where the owning type / member may change in the future to be incompatible with `allows ref struct`. For example:
+
+- An `abstract` method which may later change to a `virtual` method
+- An `abstract` type which may later add implementations
+- An `interface` which may later add a default interface method
```

Comment:
This last one is interesting. That means we're going to want to be thoughtful about which interfaces this applies to, e.g. we probably want it on `IEnumerable<T>` but probably not on `IList<T>`.

Reply:
This is similar to the existing challenges posed by DIM. Specifically [cases][1] where adding members but only re-compiling _some_ of the DLLs will lead to a runtime exception. This is effectively another instance of that. Adding a DIM will cause a runtime exception if it's ever invoked on a `ref struct` that _inheritted_ that DIM. If the code containing the `ref struct` were re-compiled the compiler would issue an error forcing an implementation that avoided the exception. 

[1]: https://github.com/dotnet/csharplang/blob/main/meetings/2018/LDM-2018-10-17.md#diamond-inheritance

Reply:
It's a shame that DIMs box on usage of `this`, I usually avoid them for that reason alone, other than with static DIMs obviously which would presumably(?) be fine with `ref struct`s.

Reply:
I don't think there are any issues with static DIMs